### ISSUE TITLE
Snippet implemented by vim function

### ISSUE THREAD

**Issue by powerman** [SirVer/ultisnips#1335]
I wonder is there any way to trigger vim function which will insert snippet?

Example use case is for Go, where https://github.com/fatih/vim-go implements support for https://github.com/koron/iferr tool by providing ways to call it using `go#iferr#Generate()` or `:GoIfErr`. This tool implements smart alternative to default snippet https://github.com/fatih/vim-go/blob/master/gosnippets/UltiSnips/go.snippets#L162-L167 which instead of fixed string `return err` inserts `return nil, err` or `return 0, "", err` etc. depending on amount and types of returned values from current function.

I've tried to use `!v go#iferr#Generate()` but it doesn't work well, because both UltiSnips and this function will make some changes/cursor movements without knowing about each other and this result in conflicts. Ideally I need a way to call this function without having UltiSnips to do anything else (no text/cursor changes).

My current workaround is to use `abbr` and trigger it with `Ctrl-]`, so when I type `errn<Tab>` I get usual snippet and when I type `errn<C-]>` I get result from that tool, but I hope there is a better way:
```vim
autocmd FileType go iabbr <buffer> errn <C-R>=<SID>goiferr()<CR>
function! s:goiferr()
    call go#iferr#Generate()
    return "\<Esc>O"
endfunction
```

**Comment by SirVer** [comment]
Yes — UltiSnips already supports calling a Vim function whose return value is the snippet body. Two ways:

1. `:help UltiSnips-anonymous-snippets` / `UltiSnips#Anon(value, ...)` from an `inoremap` or mapping. Pass the function's return value as `value`.

2. Inside a snippet body, use `` `!v func()` `` or `` `!p snip.rv = vim.eval('func()')` ``.

The Go `iferr` case in the OP is awkward because `go#iferr#Generate()` mutates the buffer directly rather than returning the text — UltiSnips can't co-operate with a function that's doing its own buffer edits mid-expansion. Wrap it to return a string (or keep your `iabbr` workaround) and it falls into the supported path above.

Closing as answered.
